app/api.js

```diff
@@ -91,47 +91,102 @@ export async function setPassword(id, owner_token, keychain) {
   return response.ok;
 }
 
-export function uploadFile(
+function asyncInitWebSocket(server) {
+  return new Promise(resolve => {
+    const ws = new WebSocket(server);
+    ws.onopen = () => {
+      resolve(ws);
+    };
+  });
+}
+
+async function upload(
+  ws,
+  stream,
+  streamInfo,
+  metadata,
+  verifierB64,
+  keychain,
+  onprogress,
+  canceller
+) {
+  const metadataHeader = arrayToB64(new Uint8Array(metadata));
+  const fileMeta = {
+    fileMetadata: metadataHeader,
+    authorization: `send-v1 ${verifierB64}`
+  };
+
+  function listenForResponse() {
+    return new Promise((resolve, reject) => {
+      ws.addEventListener('message', function(msg) {
+        const response = JSON.parse(msg.data);
+        if (response.error) {
+          reject(response.error);
+        } else {
+          resolve({
+            url: response.url,
+            id: response.id,
+            ownerToken: response.owner
+          });
+        }
+      });
+    });
+  }
+
+  const resPromise = listenForResponse();
+  ws.send(JSON.stringify(fileMeta));
+
+  const reader = stream.getReader();
+  let state = await reader.read();
+  let size = 0;
+  while (!state.done) {
+    const buf = state.value;
+    if (canceller.cancelled) {
+      ws.close(4000, 'upload cancelled');
+      throw new Error(0);
+    }
+    ws.send(buf);
+
+    onprogress([Math.min(streamInfo.fileSize, size), streamInfo.fileSize]);
+    size += streamInfo.recordSize;
+    state = await reader.read();
+  }
+
+  const response = await resPromise;
+
+  ws.close();
+  return response;
+}
+
+export async function uploadWs(
   encrypted,
+  info,
   metadata,
   verifierB64,
   keychain,
   onprogress
 ) {
-  const xhr = new XMLHttpRequest();
-  const upload = {
+  const host = window.location.hostname;
+  const port = window.location.port;
+  const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
+  const ws = await asyncInitWebSocket(`${protocol}//${host}:${port}/api/ws`);
```

Comment:
I think it makes more sense to construct the socket in `upload`. That way the lifespan of `ws` is fully contained in that function.